### ISSUE TITLE
Discuss if startSpan() and end() should be throwing

### ISSUE THREAD

**Issue by slashmo** [slashmo/gsoc-swift-tracing#96]
@ktoso & I discussed this a bit on a call already. Although there are some examples of where throwing in the `startSpan()` & `end()` methods could improve the specific tracer implementation, we think that it may be too verbose for the end-user of the instrumented system. Instead, in cases like calling `end()` on an already emitted `Span` should in my opinion log a warning and otherwise ignore the call.

Happy to continue that discussion and get real-world input from @pokryfka on this.

**Comment by pokryfka** [comment]
Fully agree that would be an extreme approach, tracer is after all "a logger" .

For the record official XRay recorders take different approach as described in https://github.com/pokryfka/aws-xray-sdk-swift/issues/28

I am thinking to create new valid context with disabled (unknown) sampling decision, which effectively would mean it will not be recorder; example:

```swift
    /// Creates new segment.
    /// Extracts the thre context from the baggage.
    /// Creates new if the baggage does not contain a valid XRay Trace Context.
    /// - Parameters:
    ///   - name: segment name
    ///   - baggage: baggage with the trace context
    ///   - metadata: segment metadata
    /// - Returns: new segment
    public func beginSegment(name: String, baggage: BaggageContext, metadata: Segment.Metadata? = nil) -> XRayRecorder.Segment {
        // TODO: Configure context missing strategy https://github.com/pokryfka/aws-xray-sdk-swift/issues/28
        let context = baggage.xRayContext ?? TraceContext(sampled: .unknown)
        return beginSegment(name: name, context: context, baggage: baggage, metadata: metadata)
    }
```

**Comment by ktoso** [comment]
Yeah the defaulting to extracting a noop span is quite good I think... Though I'll ask some amazon folks about this when I'll get a chance.

I agree with @pokryfka that that's a nice option, returning a noop span like this.

I do like that xray tracers usually allow configuring this so that's good. 

The question actual question though is about _in general_ what the API should even allow, i.e. if a tracer really wanted to do throwing, we would have to make all our APIs throwing, including injecting and extracting etc... Which honestly seems somewhat overkill, and not really like an error end-users are able to recover from anyway, thus I'm more leaning towards keeping the APIs non-throwing.

To clarify, the API has a few options here:

- 1 (current design) APIs for inject, extract, startSpan, end are `non-throwing`, tracers therefore have two options: 
  - a) log a warning if they really expected something or a span was double closed etc or 
  - b) _crash hard_ when something bad happens (fatal error and friends)
  - c) some form of "ignore and act as noop"; for inject/extract this is trivial, returning the passed in value, for starting a span it requires implementing a noop span (we could provide one if necessary) and for ending a span it's also handled by a noop span. One may combine this with sometimes logging etc.
- 2) mark inject/extract, startSpan and end all as `throws` 
  - ❌ realistically this is somewhat overkill and a huge pain for end users, because tracing already adds a lot of noise to the instrumentation; I also don't feel like there's much one could do in the `catch` for such startSpan failing... since the APIs are all taking spans or contexts the end user would then have to invent how to handle it, and they're most likely NOT in a good position to figure it out, so they'd likely just `try!` the calls since the vast majority of the time those would not be throwing; so our API would be leading people to bad swift patterns (try!).

Specifically we were thinking about what tracers will want to do when a span "double closed". It's kind of a programming error, and could end up causing one to report "not the span duration I expected" etc; so one would really want to know about these situations.

I kind of feel it is a programming error, so one may want to crash so it's easy to spot (backtrace) where and why this happened, but on the other hand doing so on production is the usual crashing the entire process which is a terrible idea, thus we're also leaning to recommend what xray and your xray impl does: some configuration that tracers should be able to either crash hard or just log the warning (note that they may have to rate throttle that logging (!)).

--- 

Long story short, I think we want to stay with the current APIs which are non-throwing, and indeed as you do in the xray impl, suggest that tracers either a) noop, or b) log or c) crash hard. The throwing APIs are not as nice in swift. (also note that while in XRay Java they _do_ throw, they do so as runtime exceptions so they're not enforced to be caught -- so it is not annoying and non-fatal, an option we do not have in swift).

Summing up: I vote we stay with non throwing APIs, sounds like that'll be okey with your impl as well @pokryfka 👍

**Comment by slashmo** [comment]
@ktoso Great summary 👍 I think we should add some recommendations as part of documentation on how to handle this and close the issue afterwards.

**Comment by pokryfka** [comment]
> Specifically we were thinking about what tracers will want to do when a span "double closed". It's kind of a programming error, and could end up causing one to report "not the span duration I expected" etc; so one would really want to know about these situations.

I was (and still am) not fully decided but ended up with public API (incl `start` and `end`) not throwing.

I am thinking to make it configurable with logging errors by default which could be changes to `assert` or `precondition` if explicitly configured to `RUNTIME_ERROR` when testing application but not intended for a default setting in production.

example: 

```swift
extension XRayRecorder.Segment {
    public func end() {
        try? end(Timestamp())
    }

   internal func end(_ timestamp: Timestamp) throws { ... }

   internal func emit() throws { ... }
}
```

**Comment by pokryfka** [comment]
@ktoso 

> I also don't feel like there's much one could do in the catch for such startSpan failing.

💯 

its good to know there is an error but I dont htink theres much user could do about it and logger/tracer should help and not make life difficult

last but not least, I think its good to provide some helpers with closures containing deferred "ends", something like https://github.com/pokryfka/aws-xray-sdk-swift/blob/master/Sources/AWSXRayRecorder/Recorder%2BHelpers.swift

no need to `end` explicitly:

```swift
recorder.segment(name: "Segment 2", context: context) { segment in
    try? segment.subsegment(name: "Subsegment 2.1") { segment in
        _ = segment.subsegment(name: "Subsegment 2.1.1 with Result") { _ -> String in
            usleep(100_000)
            return "Result"
        }
        try segment.subsegment(name: "Subsegment 2.1.2 with Error") { _ in
            usleep(200_000)
            throw ExampleError.test
        }
    }
}
```

note that it does not `throw` but it does `rethrow` so no need to `try` subsegment "Subsegment 2.1.1 "

**Comment by ktoso** [comment]
re: the edit: Sugar is a separate topic and not something we should jump on just yet, I'll make a separate ticket for it though.

Consider tracer.withSpan() { ... } convenience APIs #98

**Comment by ktoso** [comment]
Closing this thread though with the decision: the functions are _not_ throwing.
